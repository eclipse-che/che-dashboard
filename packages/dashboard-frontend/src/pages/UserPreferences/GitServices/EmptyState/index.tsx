/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { EmptyState, EmptyStateIcon, EmptyStateVariant, Title } from '@patternfly/react-core';
import { RegistryIcon } from '@patternfly/react-icons';
import React from 'react';

export type Props = {
  text: string;
};

export class GitServicesEmptyState extends React.PureComponent<Props> {
  public render(): React.ReactElement {
    return (
      <EmptyState isFullHeight={true} variant={EmptyStateVariant.small}>
        <EmptyStateIcon icon={RegistryIcon} />
        <Title headingLevel="h4" size="lg">
          {this.props.text}
        </Title>
      </EmptyState>
    );
  }
}
