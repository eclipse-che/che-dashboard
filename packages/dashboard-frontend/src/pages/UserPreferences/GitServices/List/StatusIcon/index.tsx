/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { api } from '@eclipse-che/common';
import {
  CheckCircleIcon,
  ExclamationTriangleIcon,
  ResourcesEmptyIcon,
} from '@patternfly/react-icons';
import React from 'react';

import { CheTooltip } from '@/components/CheTooltip';

export type Props = {
  gitProvider: api.GitOauthProvider;
  providersWithToken: api.GitOauthProvider[];
  skipOauthProviders: api.GitOauthProvider[];
};

export class GitServiceStatusIcon extends React.PureComponent<Props> {
  private isSkipOauth(): boolean {
    const { gitProvider, skipOauthProviders } = this.props;
    // Use includes filter to handle the bitbucket-server oauth 2 provider.
    // The bitbucket server oauth2 provider name is 'bitbucket',
    // but the corresponding 'skip oauth' item is 'bitbucket-server'.
    return skipOauthProviders.some(s => s.includes(gitProvider));
  }

  private hasOauthToken(): boolean {
    const { gitProvider, providersWithToken } = this.props;
    // Use includes filter to handle the bitbucket-server oauth 2 provider.
    // The bitbucket server oauth2 provider name is 'bitbucket',
    // but the corresponding 'provider with token' item is 'bitbucket-server'.
    return providersWithToken.some(p => p.includes(gitProvider));
  }

  public render(): React.ReactElement {
    const hasOauthToken = this.hasOauthToken();
    const isSkipOauth = this.isSkipOauth();

    if (hasOauthToken) {
      return (
        <CheTooltip content={<span>Authorized</span>}>
          <CheckCircleIcon
            data-testid="icon-authorized"
            color="var(--pf-global--success-color--100)"
          />
        </CheTooltip>
      );
    } else if (isSkipOauth) {
      return (
        <CheTooltip content={<span>Authorization has been rejected by user.</span>}>
          <ExclamationTriangleIcon
            data-testid="icon-declined"
            color="var(--pf-global--warning-color--100)"
          />
        </CheTooltip>
      );
    }

    return (
      <CheTooltip content={<span>Unauthorized</span>}>
        <ResourcesEmptyIcon
          data-testid="icon-unauthorized"
          color="var(--pf-global--disabled-color--100)"
        />
      </CheTooltip>
    );
  }
}
