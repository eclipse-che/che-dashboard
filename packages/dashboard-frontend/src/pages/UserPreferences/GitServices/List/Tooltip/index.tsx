/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { Button, Icon } from '@patternfly/react-core';
import { ExternalLinkSquareAltIcon, WarningTriangleIcon } from '@patternfly/react-icons';
import React from 'react';

import { CheTooltip } from '@/components/CheTooltip';

export type Props = {
  isVisible: boolean;
  serverURI: string;
};

export class GitServiceTooltip extends React.PureComponent<Props> {
  private handleClick(): void {
    window.open(this.props.serverURI, '_blank');
  }

  public render(): React.ReactElement {
    const { isVisible } = this.props;

    if (isVisible === false) {
      return <></>;
    }

    const content = (
      <span data-testid="tooltip-content">
        Provided API does not support the automatic token revocation. You can revoke it manually on
        &nbsp;
        <Button
          component="a"
          data-testid="tooltip-link"
          icon={<ExternalLinkSquareAltIcon />}
          iconPosition="right"
          isInline={true}
          variant="link"
          onClick={() => this.handleClick()}
        >
          {this.props.serverURI}
        </Button>
        .
      </span>
    );

    return (
      <Icon isInline={true} status="warning" data-testid="tooltip-icon">
        <CheTooltip content={content}>
          <WarningTriangleIcon />
        </CheTooltip>
      </Icon>
    );
  }
}
