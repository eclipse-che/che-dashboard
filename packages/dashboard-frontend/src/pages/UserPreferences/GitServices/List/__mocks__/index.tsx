/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import React from 'react';

import { Props } from '@/pages/UserPreferences/GitServices/List';

export class GitServicesList extends React.PureComponent<Props> {
  render(): React.ReactNode {
    const { isDisabled, gitOauth, providersWithToken, skipOauthProviders, onRevokeServices } =
      this.props;

    return (
      <div data-testid="git-services-list">
        <div>GitServicesList</div>
        <div data-testid="number-of-git-services">{gitOauth.length}</div>
        <div data-testid="providers-with-token">{providersWithToken.join()}</div>
        <div data-testid="skip-oauth-providers">{skipOauthProviders.join()}</div>
        <div data-testid="list-is-disabled">{isDisabled ? 'true' : 'false'}</div>
        <button onClick={() => onRevokeServices(gitOauth)}>Revoke</button>
      </div>
    );
  }
}
