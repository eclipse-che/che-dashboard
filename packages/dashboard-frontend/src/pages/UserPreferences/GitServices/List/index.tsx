/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import { api } from '@eclipse-che/common';
import { Button, ButtonVariant } from '@patternfly/react-core';
import {
  ActionsColumn,
  IAction,
  TableComposable,
  TableVariant,
  Tbody,
  Td,
  Th,
  Thead,
  Tr,
} from '@patternfly/react-table';
import React from 'react';

import { GIT_OAUTH_PROVIDERS } from '@/pages/UserPreferences/const';
import { GitServiceStatusIcon } from '@/pages/UserPreferences/GitServices/List/StatusIcon';
import { GitServiceTooltip } from '@/pages/UserPreferences/GitServices/List/Tooltip';
import { GitServicesToolbar } from '@/pages/UserPreferences/GitServices/Toolbar';
import { IGitOauth } from '@/store/GitOauthConfig/types';

export const CAN_REVOKE_FROM_DASHBOARD: ReadonlyArray<api.GitOauthProvider> = [
  'github',
  'github_2',
  'gitlab',
];

export type Props = {
  isDisabled: boolean;
  gitOauth: IGitOauth[];
  providersWithToken: api.GitOauthProvider[];
  skipOauthProviders: api.GitOauthProvider[];
  onRevokeServices: (services: IGitOauth[]) => void;
};

type State = {
  selectedItems: IGitOauth[];
};

export class GitServicesList extends React.PureComponent<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      selectedItems: [],
    };
  }

  private buildHeadRow(): React.ReactElement {
    return (
      <Tr>
        <Th />
        <Th dataLabel="Git Service Name Column Header">Name</Th>
        <Th dataLabel="Git Service Endpoint URL Column Header">Server</Th>
        <Th dataLabel="Git Service Authorization Status Column Header">Authorization</Th>
        <Th dataLabel="Git Service Actions Column Header" />
      </Tr>
    );
  }

  private handleSelectItem(isSelected: boolean, rowIndex: number): void {
    const { gitOauth } = this.props;

    /* c8 ignore start */
    if (rowIndex === -1) {
      // Select all (header row checked)
      const selectedItems = isSelected && gitOauth.length > 0 ? gitOauth : [];
      this.setState({ selectedItems });
      return;
    }
    /* c8 ignore stop */

    // Select single row
    const selectedItem = gitOauth[rowIndex];
    this.setState((prevState: State) => {
      return {
        selectedItems: isSelected
          ? [...prevState.selectedItems, selectedItem]
          : prevState.selectedItems.filter(item => item !== selectedItem),
      };
    });
  }

  private deselectServices(services: IGitOauth[]): void {
    const { selectedItems } = this.state;
    this.setState({
      selectedItems: selectedItems.filter(s => !services.includes(s)),
    });
  }

  private buildBodyRows(): React.ReactNode[] {
    const { isDisabled, gitOauth, providersWithToken, skipOauthProviders } = this.props;
    const { selectedItems } = this.state;

    return (
      gitOauth
        // sort by display name
        .sort((serviceA, serviceB) => {
          return GIT_OAUTH_PROVIDERS[serviceA.name].localeCompare(
            GIT_OAUTH_PROVIDERS[serviceB.name],
          );
        })
        .map((service, rowIndex) => {
          const hasWarningMessage =
            this.isRevokeEnabled(service.name) === false &&
            this.hasOauthToken(service.name) === true;

          const canRevoke = this.isRevokeEnabled(service.name) === true;
          const hasToken = this.hasOauthToken(service.name) === true;
          const rowDisabled = isDisabled || canRevoke === false || hasToken === false;

          const actionItems = this.buildActionItems(service);

          return (
            <Tr key={service.name} data-testid={service.name}>
              <Td
                dataLabel="Git Service Checkbox"
                select={{
                  rowIndex,
                  onSelect: (_event, isSelected) => this.handleSelectItem(isSelected, rowIndex),
                  isSelected: selectedItems.includes(service),
                  disable: rowDisabled,
                }}
              />
              <Td dataLabel="Git Service Name">
                {GIT_OAUTH_PROVIDERS[service.name]}{' '}
                <GitServiceTooltip isVisible={hasWarningMessage} serverURI={service.endpointUrl} />
              </Td>
              <Td dataLabel="Git Service Endpoint URL">
                <Button
                  component="a"
                  variant={ButtonVariant.link}
                  href={service.endpointUrl}
                  isInline={true}
                  target="_blank"
                  rel="noreferer"
                >
                  {service.endpointUrl}
                </Button>
              </Td>
              <Td dataLabel="Git Service Authorization">
                <GitServiceStatusIcon
                  gitProvider={service.name}
                  providersWithToken={providersWithToken}
                  skipOauthProviders={skipOauthProviders}
                />
              </Td>
              <Td dataLabel="Git Service Actions" isActionCell={true}>
                <ActionsColumn isDisabled={rowDisabled} items={actionItems} />
              </Td>
            </Tr>
          );
        })
    );
  }

  private buildActionItems(service: IGitOauth): IAction[] {
    return [
      {
        title: 'Revoke',
        onClick: () => this.handleRevokeService(service),
      },
    ];
  }

  private isRevokeEnabled(providerName: api.GitOauthProvider): boolean {
    return CAN_REVOKE_FROM_DASHBOARD.includes(providerName) === true;
  }

  private hasOauthToken(providerName: api.GitOauthProvider): boolean {
    return this.props.providersWithToken.includes(providerName);
  }

  private handleRevokeService(service: IGitOauth): void {
    this.props.onRevokeServices([service]);
    this.deselectServices([service]);
  }

  private async handleRevokeSelectedServices(): Promise<void> {
    const { selectedItems } = this.state;

    this.props.onRevokeServices(selectedItems);
    this.deselectServices(selectedItems);
  }

  render(): React.ReactNode {
    const { isDisabled } = this.props;
    const { selectedItems } = this.state;

    const headRow = this.buildHeadRow();
    const bodyRows = this.buildBodyRows();

    return (
      <React.Fragment>
        <GitServicesToolbar
          isDisabled={isDisabled}
          selectedItems={selectedItems}
          onRevokeButton={async () => await this.handleRevokeSelectedServices()}
        />

        <TableComposable aria-label="Git Services" variant={TableVariant.compact}>
          <Thead>{headRow}</Thead>
          <Tbody>{bodyRows}</Tbody>
        </TableComposable>
      </React.Fragment>
    );
  }
}
