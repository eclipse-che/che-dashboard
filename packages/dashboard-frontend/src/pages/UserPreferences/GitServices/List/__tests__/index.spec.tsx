/*
 * Copyright (c) 2018-2024 Red Hat, Inc.
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *   Red Hat, Inc. - initial API and implementation
 */

import userEvent from '@testing-library/user-event';
import React from 'react';

import { GitServicesList, Props } from '@/pages/UserPreferences/GitServices/List';
import getComponentRenderer, { screen, within } from '@/services/__mocks__/getComponentRenderer';

jest.mock('@/pages/UserPreferences/GitServices/List/StatusIcon');
jest.mock('@/pages/UserPreferences/GitServices/List/Tooltip');
jest.mock('@/pages/UserPreferences/GitServices/Toolbar');

const { createSnapshot, renderComponent } = getComponentRenderer(getComponent);

describe('GitServicesList', () => {
  let props: Props;

  beforeEach(() => {
    props = {
      gitOauth: [
        {
          name: 'github',
          endpointUrl: 'https://github.com',
        },
        {
          name: 'gitlab',
          endpointUrl: 'https://gitlab.com',
        },
        {
          name: 'bitbucket',
          endpointUrl: 'https://bitbucket.com',
        },
      ],
      isDisabled: false,
      onRevokeServices: jest.fn(),
      onClearService: jest.fn(),
      providersWithToken: ['github', 'gitlab'],
      skipOauthProviders: [],
    };
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  test('snapshot', () => {
    const snapshot = createSnapshot(props);
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test('initial state', () => {
    renderComponent(props);

    const rows = screen.getAllByRole('row');
    expect(rows).toHaveLength(4);

    // The very first row in the table is the header, skip it

    // The next row is Bitbucket
    {
      const bitbucketRow = rows[1];
      expect(bitbucketRow).toHaveTextContent('bitbucket');
      expect(bitbucketRow).toHaveTextContent('https://bitbucket.com');

      const checkbox = within(bitbucketRow).getByRole('checkbox');
      expect(checkbox).toBeDisabled();
      expect(checkbox).not.toBeChecked();

      const kebab = within(bitbucketRow).getByRole('button', { name: 'Actions' });
      expect(kebab).toBeDisabled();
    }

    // The next row is Github
    {
      const githubRow = rows[2];
      expect(githubRow).toHaveTextContent('github');
      expect(githubRow).toHaveTextContent('https://github.com');

      const checkbox = within(githubRow).getByRole('checkbox');
      expect(checkbox).toBeEnabled();
      expect(checkbox).not.toBeChecked();

      const kebab = within(githubRow).getByRole('button', { name: 'Actions' });
      expect(kebab).toBeEnabled();
    }

    // The next row is Gitlab
    {
      const gitlabRow = rows[3];
      expect(gitlabRow).toHaveTextContent('gitlab');
      expect(gitlabRow).toHaveTextContent('https://gitlab.com');

      const checkbox = within(gitlabRow).getByRole('checkbox');
      expect(checkbox).toBeEnabled();
      expect(checkbox).not.toBeChecked();

      const kebab = within(gitlabRow).getByRole('button', { name: 'Actions' });
      expect(kebab).toBeEnabled();
    }
  });

  test('service revocable (gitlab)', () => {
    renderComponent(props);

    const rows = screen.getAllByRole('row');

    // get the github row
    const gitlabRow = rows[3];
    expect(gitlabRow).toHaveTextContent('gitlab');

    const gitlabCheckbox = within(gitlabRow).getByRole('checkbox');
    const gitlabKebab = within(gitlabRow).getByRole('button', { name: 'Actions' });

    expect(gitlabCheckbox).toBeEnabled();
    expect(gitlabCheckbox).not.toBeChecked();

    expect(gitlabKebab).toBeEnabled();
  });

  test('service revocable (github)', async () => {
    renderComponent(props);

    const rows = screen.getAllByRole('row');

    // get the github row
    const githubRow = rows[2];
    expect(githubRow).toHaveTextContent('github');

    const githubCheckbox = within(githubRow).getByRole('checkbox');
    const githubKebab = within(githubRow).getByRole('button', { name: 'Actions' });

    // the checkbox is enabled and unchecked
    expect(githubCheckbox).toBeEnabled();
    expect(githubCheckbox).not.toBeChecked();

    // check the checkbox
    await userEvent.click(githubCheckbox);
    expect(githubCheckbox).toBeChecked();

    // uncheck the checkbox
    await userEvent.click(githubCheckbox);
    expect(githubCheckbox).not.toBeChecked();

    // the kebab button is enabled
    expect(githubKebab).toBeEnabled();

    // revoke button is not present
    {
      const revokeButton = within(githubRow).queryByRole('menuitem', { name: 'Revoke' });
      expect(revokeButton).toBeNull();
    }

    // open kebab menu
    await userEvent.click(githubKebab);

    // the revoke button is present
    const revokeButton = within(githubRow).queryByRole('menuitem', { name: 'Revoke' });
    expect(revokeButton).not.toBeNull();

    // click the revoke button
    await userEvent.click(revokeButton!);

    expect(props.onRevokeServices).toHaveBeenCalledTimes(1);
    expect(props.onRevokeServices).toHaveBeenCalledWith([
      {
        name: 'github',
        endpointUrl: 'https://github.com',
      },
    ]);
  });

  test('can clear opt-out (github)', async () => {
    props = {
      gitOauth: [
        {
          name: 'github',
          endpointUrl: 'https://github.com',
        },
      ],
      isDisabled: false,
      onRevokeServices: jest.fn(),
      onClearService: jest.fn(),
      providersWithToken: [],
      skipOauthProviders: ['github'],
    };
    renderComponent(props);

    const rows = screen.getAllByRole('row');

    // get the github row
    const githubRow = rows[1];
    expect(githubRow).toHaveTextContent('github');

    const githubCheckbox = within(githubRow).getByRole('checkbox');
    const githubKebab = within(githubRow).getByRole('button', { name: 'Actions' });

    // the checkbox is disabled and unchecked
    expect(githubCheckbox).toBeDisabled();
    expect(githubCheckbox).not.toBeChecked();

    // the kebab button is enabled
    expect(githubKebab).toBeEnabled();

    // Clear button is not present
    {
      const clearButton = within(githubRow).queryByRole('menuitem', { name: 'Clear' });
      expect(clearButton).toBeNull();
    }

    // open kebab menu
    await userEvent.click(githubKebab);

    // the Clear button is present
    const clearButton = within(githubRow).queryByRole('menuitem', { name: 'Clear' });
    expect(clearButton).not.toBeNull();

    // click the Clear button
    await userEvent.click(clearButton!);

    expect(props.onClearService).toHaveBeenCalledTimes(1);
    expect(props.onClearService).toHaveBeenCalledWith('github');
  });

  test('toolbar', async () => {
    renderComponent(props);

    const selectedItemsCount = screen.getByTestId('selected-items-count');

    // No items selected
    expect(selectedItemsCount).toHaveTextContent('0');

    const rows = screen.getAllByRole('row');

    // check the github row
    const githubRow = rows[2];
    const githubCheckbox = within(githubRow).getByRole('checkbox');
    await userEvent.click(githubCheckbox);

    // One item selected
    expect(selectedItemsCount).toHaveTextContent('1');

    // the revoke button in toolbar
    const revokeButton = screen.getByRole('button', { name: 'Revoke' });
    await userEvent.click(revokeButton);

    expect(props.onRevokeServices).toHaveBeenCalledTimes(1);
    expect(props.onRevokeServices).toHaveBeenCalledWith([
      {
        name: 'github',
        endpointUrl: 'https://github.com',
      },
    ]);
  });

  test('disabled list', () => {
    // initially prop isDisabled is false
    const { reRenderComponent } = renderComponent(props);

    const rows = screen.getAllByRole('row');

    // get the github row
    const githubRow = rows[2];
    expect(githubRow).toHaveTextContent('github');

    // get the github row controls
    const githubCheckbox = within(githubRow).getByRole('checkbox');
    const githubKebab = within(githubRow).getByRole('button', { name: 'Actions' });

    // they are enabled
    expect(githubCheckbox).toBeEnabled();
    expect(githubKebab).toBeEnabled();

    // toolbar is enabled
    const revokeButtonDisabled = screen.getByTestId('revoke-is-disabled');
    expect(revokeButtonDisabled).toHaveTextContent('false');

    // set isDisabled to be true
    reRenderComponent({ ...props, isDisabled: true });

    // the github row controls are disabled
    expect(githubCheckbox).toBeDisabled();

    // toolbar is disabled
    expect(revokeButtonDisabled).toHaveTextContent('true');
  });
});

function getComponent(props: Props): React.ReactElement<Props> {
  return (
    <GitServicesList
      gitOauth={props.gitOauth}
      isDisabled={props.isDisabled}
      onRevokeServices={props.onRevokeServices}
      onClearService={props.onClearService}
      providersWithToken={props.providersWithToken}
      skipOauthProviders={props.skipOauthProviders}
    />
  );
}
