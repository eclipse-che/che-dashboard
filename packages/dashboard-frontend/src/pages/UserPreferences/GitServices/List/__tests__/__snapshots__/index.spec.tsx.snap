// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GitServicesList snapshot 1`] = `
[
  <div
    data-testid="git-services-toolbar"
  >
    <div>
      GitServicesToolbar
    </div>
    <div
      data-testid="revoke-is-disabled"
    >
      false
    </div>
    <div
      data-testid="selected-items-count"
    >
      0
    </div>
    <button
      onClick={[Function]}
    >
      Revoke
    </button>
  </div>,
  <table
    aria-label="Git Services"
    className="pf-c-table pf-m-grid-md pf-m-compact"
    data-ouia-component-id="OUIA-Generated-Table-1"
    data-ouia-component-type="PF4/Table"
    data-ouia-safe={true}
    role="grid"
  >
    <thead
      className=""
    >
      <tr
        aria-label=""
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-1"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        hidden={false}
      >
        <th
          className=""
          onMouseEnter={[Function]}
          scope={null}
        />
        <th
          className=""
          data-label="Git Service Name Column Header"
          onMouseEnter={[Function]}
          scope="col"
        >
          Name
        </th>
        <th
          className=""
          data-label="Git Service Endpoint URL Column Header"
          onMouseEnter={[Function]}
          scope="col"
        >
          Server
        </th>
        <th
          className=""
          data-label="Git Service Authorization Status Column Header"
          onMouseEnter={[Function]}
          scope="col"
        >
          Authorization
        </th>
        <th
          className=""
          data-label="Git Service Actions Column Header"
          onMouseEnter={[Function]}
          scope={null}
        />
      </tr>
    </thead>
    <tbody
      className=""
      role="rowgroup"
    >
      <tr
        aria-label=""
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-2"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        data-testid="bitbucket"
        hidden={false}
      >
        <td
          className="pf-c-table__check"
          data-label="Git Service Checkbox"
          onMouseEnter={[Function]}
        >
          <label>
            <input
              aria-label="Select row 0"
              checked={false}
              disabled={true}
              name="checkrow0"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </td>
        <td
          className=""
          data-label="Git Service Name"
          onMouseEnter={[Function]}
        >
          Bitbucket
           
          <span
            data-testid="git-service-tooltip"
          >
            <span>
              isVisible: 
              false
            </span>
          </span>
        </td>
        <td
          className=""
          data-label="Git Service Endpoint URL"
          onMouseEnter={[Function]}
        >
          <a
            aria-disabled={false}
            aria-label={null}
            className="pf-c-button pf-m-link pf-m-inline"
            data-ouia-component-id="OUIA-Generated-Button-link-1"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            disabled={null}
            href="https://bitbucket.com"
            rel="noreferer"
            role={null}
            target="_blank"
            type={null}
          >
            https://bitbucket.com
          </a>
        </td>
        <td
          className=""
          data-label="Git Service Authorization"
          onMouseEnter={[Function]}
        >
          <span
            data-testid="git-service-status-icon"
          >
            <span>
              Authentication Status Icon for 
              bitbucket
            </span>
          </span>
        </td>
        <td
          className="pf-c-table__action"
          data-label="Git Service Actions"
          onMouseEnter={[Function]}
        >
          <div
            className="pf-c-dropdown pf-m-align-right"
            data-ouia-component-id="OUIA-Generated-Dropdown-1"
            data-ouia-component-type="PF4/Dropdown"
            data-ouia-safe={true}
          >
            <button
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="Actions"
              className="pf-c-dropdown__toggle pf-m-plain"
              disabled={true}
              id="pf-dropdown-toggle-id-0"
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 192 512"
                width="1em"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
      <tr
        aria-label=""
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-3"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        data-testid="github"
        hidden={false}
      >
        <td
          className="pf-c-table__check"
          data-label="Git Service Checkbox"
          onMouseEnter={[Function]}
        >
          <label>
            <input
              aria-label="Select row 1"
              checked={false}
              name="checkrow1"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </td>
        <td
          className=""
          data-label="Git Service Name"
          onMouseEnter={[Function]}
        >
          GitHub
           
          <span
            data-testid="git-service-tooltip"
          >
            <span>
              isVisible: 
              false
            </span>
          </span>
        </td>
        <td
          className=""
          data-label="Git Service Endpoint URL"
          onMouseEnter={[Function]}
        >
          <a
            aria-disabled={false}
            aria-label={null}
            className="pf-c-button pf-m-link pf-m-inline"
            data-ouia-component-id="OUIA-Generated-Button-link-2"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            disabled={null}
            href="https://github.com"
            rel="noreferer"
            role={null}
            target="_blank"
            type={null}
          >
            https://github.com
          </a>
        </td>
        <td
          className=""
          data-label="Git Service Authorization"
          onMouseEnter={[Function]}
        >
          <span
            data-testid="git-service-status-icon"
          >
            <span>
              Authentication Status Icon for 
              github
            </span>
          </span>
        </td>
        <td
          className="pf-c-table__action"
          data-label="Git Service Actions"
          onMouseEnter={[Function]}
        >
          <div
            className="pf-c-dropdown pf-m-align-right"
            data-ouia-component-id="OUIA-Generated-Dropdown-2"
            data-ouia-component-type="PF4/Dropdown"
            data-ouia-safe={true}
          >
            <button
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="Actions"
              className="pf-c-dropdown__toggle pf-m-plain"
              disabled={false}
              id="pf-dropdown-toggle-id-1"
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 192 512"
                width="1em"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
      <tr
        aria-label=""
        className=""
        data-ouia-component-id="OUIA-Generated-TableRow-4"
        data-ouia-component-type="PF4/TableRow"
        data-ouia-safe={true}
        data-testid="gitlab"
        hidden={false}
      >
        <td
          className="pf-c-table__check"
          data-label="Git Service Checkbox"
          onMouseEnter={[Function]}
        >
          <label>
            <input
              aria-label="Select row 2"
              checked={false}
              name="checkrow2"
              onChange={[Function]}
              type="checkbox"
            />
          </label>
        </td>
        <td
          className=""
          data-label="Git Service Name"
          onMouseEnter={[Function]}
        >
          GitLab
           
          <span
            data-testid="git-service-tooltip"
          >
            <span>
              isVisible: 
              false
            </span>
          </span>
        </td>
        <td
          className=""
          data-label="Git Service Endpoint URL"
          onMouseEnter={[Function]}
        >
          <a
            aria-disabled={false}
            aria-label={null}
            className="pf-c-button pf-m-link pf-m-inline"
            data-ouia-component-id="OUIA-Generated-Button-link-3"
            data-ouia-component-type="PF4/Button"
            data-ouia-safe={true}
            disabled={null}
            href="https://gitlab.com"
            rel="noreferer"
            role={null}
            target="_blank"
            type={null}
          >
            https://gitlab.com
          </a>
        </td>
        <td
          className=""
          data-label="Git Service Authorization"
          onMouseEnter={[Function]}
        >
          <span
            data-testid="git-service-status-icon"
          >
            <span>
              Authentication Status Icon for 
              gitlab
            </span>
          </span>
        </td>
        <td
          className="pf-c-table__action"
          data-label="Git Service Actions"
          onMouseEnter={[Function]}
        >
          <div
            className="pf-c-dropdown pf-m-align-right"
            data-ouia-component-id="OUIA-Generated-Dropdown-3"
            data-ouia-component-type="PF4/Dropdown"
            data-ouia-safe={true}
          >
            <button
              aria-expanded={false}
              aria-haspopup={true}
              aria-label="Actions"
              className="pf-c-dropdown__toggle pf-m-plain"
              disabled={false}
              id="pf-dropdown-toggle-id-2"
              onClick={[Function]}
              onKeyDown={[Function]}
              type="button"
            >
              <svg
                aria-hidden={true}
                aria-labelledby={null}
                fill="currentColor"
                height="1em"
                role="img"
                style={
                  {
                    "verticalAlign": "-0.125em",
                  }
                }
                viewBox="0 0 192 512"
                width="1em"
              >
                <path
                  d="M96 184c39.8 0 72 32.2 72 72s-32.2 72-72 72-72-32.2-72-72 32.2-72 72-72zM24 80c0 39.8 32.2 72 72 72s72-32.2 72-72S135.8 8 96 8 24 40.2 24 80zm0 352c0 39.8 32.2 72 72 72s72-32.2 72-72-32.2-72-72-72-72 32.2-72 72z"
                />
              </svg>
            </button>
          </div>
        </td>
      </tr>
    </tbody>
  </table>,
]
`;
